### ISSUE TITLE
Access to the HA states object in an EVAL function

### ISSUE THREAD

**Issue by parautenbach** [RomRider/apexcharts-card#615]
**Checklist**
- [x] I am running the latest version of the card
- [x] I checked if there isn't another feature request opened with the same request
- [x] I checked that the feature is not already available in the `dev` branch

**Is your feature request related to a problem? Please describe.**
I have a highly dynamic chart. My original idea was to allow and EVAL function to set the min and max of a radialBar chart, but I've since figured I don't need that, so instead I'd like to access another entity's value from within a `formatter`'s `EVAL: function` block. 

My example: I have a radialBar chart that shows the percentage of fair weather for the current day. My HA template sensor doesn't take clear skies at night into account, since I'm only interested in daytime fair weather. The consequence is that the percentage of fair weather time per day is then a function of the daylength, which is dynamic. In other words, if the daylength today is 13h15 and the fair weather was 10h12, then it's 76.98%, but tomorrow the daylength will be different. I'd like to use an EVAL function then to access my daylength sensor in order to set the `max` field for the radialBar chart. 

**Describe the solution you'd like**
A `states` object with access to all entities. Many custom cards, e.g., the famous custom button card and card-mod allow this. 

**Describe alternatives you've considered**
I've tried a number of hacks, but nothing that could achieve this.

**Additional context**
Nothing I can think of.

**Comment by parautenbach** [comment]
.

**Comment by will2022** [comment]
I think I might have a similar feature request. It would be nice if I could show or hide the entity in the graph depending on the value of an entity state. I'm unsure if there's a way to achieve this.
```yaml
series:
  - entity: sensor.some_sensor
    show:
      in_chart: '{{state_attr("sensor.some_sensor", "boolean_state")}}'```

**Comment by parautenbach** [comment]
It would still be nice to get more dynamic bahaviour. :-)

**Comment by parautenbach** [comment]
.

**Comment by jorritjaap** [comment]
@parautenbach 
As an alternative I am using the show object to hide the senor from the chart, but use it in the fucntion.

`type: custom:apexcharts-card
series:
  - entity: sensor.solar_house_consumption_w
    show:
      in_chart: false
`

Maybe that can work for you use case too?

**Comment by parautenbach** [comment]
Thanks, but that's not quite what's needed here. 

Perhaps a picture would help:

<img width="252" alt="Screenshot 2024-06-04 at 10 22 51" src="https://github.com/RomRider/apexcharts-card/assets/743320/20b8d373-f8e2-463c-b323-e19df2f288f0">

Here I'm displaying both a percentage of the day that's fair weather as well as the duration. Day length obviously varies over a year, so the problem is that what's 100% changes. 

I managed to get this working at the time, but completely forgot to circle back here.

This [post](https://community.home-assistant.io/t/apexcharts-card-a-highly-customizable-graph-card/272877/2608) on the community forum helped.

This is my current solution, which is fine, but can definitely be more readable and digest with direct access to the states object:

```
              - type: custom:apexcharts-card
                series:
                  - entity: sensor.fair_weather_time
                    transform: "return x / parseFloat(hass.states['sensor.home_sun_daylight'].state) * 100;"
                    color: var(--custom-color-yellow)
                    min: 0
                    max: 100
                chart_type: radialBar
                apex_config:
                  title:
                    text: Fair Weather
                  legend:
                    show: false
                  plotOptions:
                    radialBar:
                      startAngle: -170
                      endAngle: 170
                      dataLabels:
                        name:
                          show: false
                        value:
                          show: true
                          fontSize: 10px
                          offsetY: -4
                          # https://community.home-assistant.io/t/apexcharts-card-a-highly-customizable-graph-card/272877/2608
                          # alternative chart: https://github.com/AmoebeLabs/flex-horseshoe-card
                          formatter: |
                            EVAL:function (percentage) {
                              // would've loved to make this snippet more readible, but getElementsByTagName returns some kind of live object
                              // which means you need to resolve the whole thing; can't make no hass object first.
                              const daylight = parseFloat(document.getElementsByTagName('home-assistant')[0].hass.states['sensor.home_sun_daylight'].state);
                              const value = percentage / 100 * parseFloat(daylight);
                              const hour = Math.floor(value);
                              const minute = Math.round((value - hour) * 60);
                              return [`${hour}h${minute.toString().padStart(2, '0')}m`, `(${percentage.toFixed(0)}%)`];
```

**Comment by parautenbach** [comment]
Please. :-)

**Comment by parautenbach** [comment]
.

**Comment by parautenbach** [comment]
.

**Comment by stalniy** [comment]
I can add access to `hass` object inside `EVAL:` functions, this will allow us to read any state from `hass` in this methods. For example, I drawing chart of instantaneous power measurements but in legend I'd like to see the total kWh during this day. I could read value of another entity which contains this details and use `legend.formatter`:

```yaml
  legend:
    formatter: |
      EVAL:function f(seriesName, opts) {
         let power = 0;
         console.log(typeof hass);
         return [
           seriesName, " - ", 
           (power * 5 / 3600 / 1000).toFixed(2) + 'kWh'];
      }
```

if @RomRider is ok with such implementation, I can send PR during next few days

**Comment by parautenbach** [comment]
@stalniy are you still happy to push your PR?

**Comment by stalniy** [comment]
Yes, I can do it if repo author approves that idea. Don’t want to make something what eventually will be ignored or thrown away

**Comment by Shultc** [comment]
I hope this will be implemented. For example I can't find a way to dymanicly hide one series from graph and make it not affect y axis bounds. But using hass in transform I will be able to just return "null" based on boolean from HA

**Comment by vingerha** [comment]
@stalniy I created a copy of this repo as more and more things bugged me with this one getting outdated. I have to be clear that I really hope that this repo will revive at some point but ... well... romrider has been dead-silent in this report for a year now.
Anyhow, I am not 100% sure if I understand the benefit and I am also not a seasoned dev but I am also happy to try things out.... so if you want, you could post a PR here based on my updated (!) version and we could let it loose on the user(s0 :)
https://github.com/vingerha/apexcharts-card

**Comment by vingerha** [comment]
> I'd be happy to help with testing.

I am advancing my repo/copy with solving multiple other issues but this one with a offered solution by @stalniy does not progress. I donot have the detailed knowledge nor the time to search for a solution and hope @stalniy can either point to a solution but even more... hopefully provide a PR

**Comment by parautenbach** [comment]
.

**Comment by parautenbach** [comment]
.

**Comment by parautenbach** [comment]
.

**Comment by parautenbach** [comment]
Still hopeful.

**Comment by parautenbach** [comment]
.

**Comment by Budgiebrain994** [comment]
It's a shame there hasn't been any action on this repo for a while now.

**Comment by thowiegit** [comment]
you might have a look at: https://github.com/dewi-ny-je/apexcharts-card

**Comment by parautenbach** [comment]
Not sure which fork is the better bet:

* https://github.com/vingerha/apexcharts-card/issues/18
* https://github.com/dewi-ny-je/apexcharts-card/issues/10

**Comment by thowiegit** [comment]
the dewi-ny-je fork should include the improvements from vingerha and has more recent activity
besides that vingerha seems to have moved to another solution for his needs, so i do not see more activity coming from him on his fork https://github.com/vingerha/apexcharts-card/issues/17 whilst dewi-ny-je seems still interested in maintaining his own fork

**Comment by dewi-ny-je** [comment]
Hi guys, I'm replying here because the discussion has been going on for a while and more people will be notified, rather than replying in the ticket https://github.com/dewi-ny-je/apexcharts-card/issues/10 of my own fork.

This is a summary evaluation by Claude Opus 5/High about the feature request. Worth noting the caveats about the implementation.

I'm open to implementing some of it to bring it in line with the other user JS-hooks, but as suggested by Claude, maybe it's worth _limiting the implementation to the minimum one_?

---

## What I read

- **Fork ticket** [dewi-ny-je#10](https://github.com/dewi-ny-je/apexcharts-card/issues/10) — asks that `hass` be passed into the scope of every `EVAL:` function, the way `transform:` already receives it.
- **Upstream** [RomRider#615](https://github.com/RomRider/apexcharts-card/issues/615) (Oct 2023, closed as stale) — same request, phrased as "a `states` object with access to all entities, like custom button-card and card-mod."

One thing directly answering your comment on the ticket: **there is no PR to link.** A search of upstream's PRs for `hass`/`eval` returns zero open or closed matches — stalniy *offered* to open one twice and never did. Caveat on sourcing: GitHub access in this session is scoped to your fork, and unauthenticated reads of #615's comment thread were blocked, so the "offered twice" detail comes from the ticket text, not from my reading the thread. Either way, this would be a from-scratch change.

## Where the code stands today

`evalApexConfig` at `src/apex-layouts.ts:553-565` is the whole feature:

```ts
const eval2 = eval;                       // indirect eval → global scope, no locals visible
apexConfig[key] = eval2(`(${...slice(5)})`);   // mutates the config in place
```

Three properties of that matter for this request:

1. It runs **indirect eval at global scope**, so the function body sees only globals — hence the `document.getElementsByTagName('home-assistant')[0].hass` workaround in the ticket.
2. It **mutates `config.apex_config` in place**, replacing the `EVAL:` string with the resulting value. The string is gone after the first pass.
3. Its only callers are `getLayoutConfig` (`apex-layouts.ts:115`) and `getBrushLayoutConfig` (`:188`), which run **once**, from `_initialLoad` (`apexcharts-card.ts:800` and `:812`).

So each `EVAL:` is evaluated exactly once per card lifetime; ApexCharts then holds the resulting function and calls it on every render. That single fact drives most of the analysis below.

For comparison, `transform` gets `('x', 'hass', 'entity')` at `graphEntry.ts:429` and `data_generator` gets `('entity', 'start', 'end', 'hass', 'moment')` at `graphEntry.ts:456`, both re-created and re-invoked on every data update — which is why they never have a staleness problem and `EVAL:` would.

## Pros

- **Consistency.** Two of the three user-JS hooks already receive `hass`. The third not doing so is an accident of implementation, not a design choice, and it's hard to explain in the YAML.
- **No new attack surface.** `EVAL:` already executes arbitrary JS with full DOM access, and the README already carries "no safeguard, use at your own risk." Everything reachable via `hass` — `callService`, `auth.data.access_token` — is already reachable through the documented DOM workaround. This makes an existing capability convenient, not new.
- **Removes a fragile idiom.** The DOM lookup depends on a `home-assistant` element being the first of its tag in the top-level document; it is unreadable in YAML and silently breaks in any context where that assumption doesn't hold.
- **Small, contained diff.** One function plus a threaded parameter through two exported signatures and two call sites.
- **Backwards compatible.** `hass` becomes a new free variable name in a scope that previously had none. It only "collides" if a user's snippet declares its own `hass`, in which case their declaration shadows the parameter and behaves exactly as before.
- **Real use cases.** Scaling a `radialBar` `max` against a daylength sensor and showing a related total in a legend formatter are both reasonable and have no clean alternative today.

## Cons and risks

**1. Staleness — this is the real one.** Giving the function access to `hass` does *not* make the chart react to that entity. The `set hass` setter (`apexcharts-card.ts:236`) only marks `updated = true` when an entity belonging to a **configured series** changes; nothing else triggers `_updateData()` (short of `update_interval`). A formatter reading `sensor.home_sun_daylight`, which is not a series, renders correctly on load and then never refreshes until something else moves the chart. Expect bug reports of the form "hass in EVAL doesn't update."

There is a clean workaround worth documenting rather than engineering around: add the entity as a series with `show.in_chart: false`. Such series stay in `_config.series` (so they drive the update path) but are excluded from `series_in_graph` at `:466`, so they don't touch the y-axis bounds. That also covers the ticket's third use case.

**2. The naive implementation is silently wrong.** Because of the mutate-once behaviour at `:558`, capturing `hass` by value at eval time binds the object present at initial load *permanently* — every later state change is invisible. The function must resolve `hass` at call time, e.g.:

```ts
function evalApexConfig(apexConfig: any, getHass: () => HomeAssistant | undefined): any {
  Object.keys(apexConfig).forEach((key) => {
    const val = apexConfig[key];
    if (typeof val === 'string' && val.trim().startsWith('EVAL:')) {
      const factory = new Function('hass', `return (${val.trim().slice(5)});`);
      const eager = factory(getHass());
      apexConfig[key] =
        typeof eager === 'function'
          ? function (this: unknown, ...args: unknown[]) {
              return factory(getHass()).apply(this, args);
            }
          : eager;
    } else if (typeof val === 'object' && val !== null) {
      apexConfig[key] = evalApexConfig(val, getHass);
    }
  });
  return apexConfig;
}
```

**3. `EVAL:` is not only used for functions.** It can produce numbers, arrays, objects. A wrapper must type-check the result, as sketched above, or it will break configs that use `EVAL:` for plain values. The eager probe that enables that check costs one extra evaluation, so an `EVAL:` snippet with side effects would run them twice — an edge case, but worth a line in the changelog.

**4. Swapping `eval` for `new Function` shifts semantics slightly.** Both evaluate in global scope, so the change is near-invisible; but do **not** add `'use strict'` here the way `transform` does. Existing snippets that assign to undeclared variables or use other sloppy-mode constructs would start throwing.

**5. Minor perf.** One closure allocation per formatter call. Formatters run per data point, so keep `new Function` outside the wrapper (as above); parsing per call would be the mistake, allocation is not.

**6. Naming.** Upstream asked for `states`, the fork ticket asks for `hass`. I'd expose `hass` alone, matching `transform` and `data_generator` — a second alias is a second thing to document and keep working forever, and `hass.states` is one property away.

**7. Scope creep in the ticket.** Use case 3 ("conditionally excluding a series based on an entity state") is not an `apex_config` formatter concern at all; it needs templating on `series[].show.in_chart`, which is a much larger feature. Worth answering explicitly so the ticket doesn't drift.

**8. Divergence.** Upstream never took this, so it becomes fork-maintained surface — though this fork already diverges substantially.

## Complexity

**Core feature: low — about half a day.**

- `src/apex-layouts.ts`: rewrite `evalApexConfig` (~20 lines), add the getter parameter to `getLayoutConfig` and `getBrushLayoutConfig`. Both already take `hass`, but as a value; they need `() => this._hass` instead, or in addition.
- `src/apexcharts-card.ts`: two call sites, `:800` and `:812`.
- README: extend the `EVAL:` section (~15 lines) with an example, the staleness caveat, and the `show.in_chart: false` recipe.
- No config-schema work — `apex_config` is untyped passthrough, so `types-config.ts` / `types-config-ti.ts` are untouched.
- Testing is entirely manual: `test/` holds only `ui-lovelace.yaml` and `configuration.yaml` fixtures, there is no unit suite. Budget most of the time here — verify a formatter, a non-function `EVAL:`, and that the value tracks a series-entity update.

Call it 1–2 hours of code, the rest documentation and hand-verification in a live HA instance.

**If reactivity is in scope: medium-to-high — 1–2 days, plus permanent API surface.** You cannot statically determine which entities an arbitrary JS body touches, so auto-refresh requires an explicit new config key (`extra_entities:` or similar), config validation in `types-config.ts` with `types-config-ti.ts` regenerated via `npm run build:types-check`, and wiring into the `set hass` update path.

My recommendation: implement the small version, document the staleness plainly alongside the hidden-series workaround, and decline the reactivity piece for now — the workaround covers it at zero maintenance cost.

Sources: [dewi-ny-je#10](https://github.com/dewi-ny-je/apexcharts-card/issues/10), [RomRider#615](https://github.com/RomRider/apexcharts-card/issues/615), [upstream PR search](https://github.com/RomRider/apexcharts-card/pulls)

**Comment by dewi-ny-je** [comment]
> the dewi-ny-je fork should include the improvements from vingerha and has more recent activity besides that vingerha seems to have moved to another solution for his needs, so i do not see more activity coming from him on his fork [vingerha/apexcharts-card#17](https://github.com/vingerha/apexcharts-card/issues/17) whilst dewi-ny-je seems still interested in maintaining his own fork

Yes I merged together all the improvements by vingerha and RomRider since the two diverged.
In fact, vingerha started using my fork. I wouldn't expect any development there, he even asked me to implement something :)

I'm not going to go to RomRider's repo and collect bug reports to fix, but if someone proposes new ideas which I find meaningful and useful I'm happy to merge them.

**Comment by Budgiebrain994** [comment]
Thanks for the forks. For what it's worth, I solved my issue with `config-template-card`.

**Comment by dewi-ny-je** [comment]
Anyone interested in testing?
https://github.com/dewi-ny-je/apexcharts-card/pull/11

**Comment by dewi-ny-je** [comment]
Released : https://github.com/dewi-ny-je/apexcharts-card/releases/tag/v2.5.0
It's quite faster in rendering, too, tahnks to the update to latest apexcharts 6.8

**Comment by Budgiebrain994** [comment]
@dewi-ny-je 

Thanks for the link. I've switched to your fork and will give it a spin. Much faster so far.

The original issue wasn't applicable to me so maybe someone else can test that specifically for you.
